modules/gapi/include/opencv2/gapi/rmat.hpp

```diff
@@ -0,0 +1,127 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+#ifndef OPENCV_GAPI_RMAT_HPP
+#define OPENCV_GAPI_RMAT_HPP
+
+#include <opencv2/gapi/gmat.hpp>
+
+namespace cv {
+
+// A lightweight wrapper on image data:
+// - Doesn't own the memory;
+// - Doesn't implement copy semantics (it's assumed that a view is created each time
+// wrapped data is being accessed);
+// - Has an optional callback which is called when the view is destroyed.
+class MatView
+{
+public:
+    using DestroyCallback = std::function<void()>;
+
+    MatView() = default;
+    MatView(const GMatDesc& desc, uchar* data, size_t step = 0u, DestroyCallback cb = nullptr)
+        : m_desc(desc), m_data(data), m_step(step == 0u ? elemSize()*cols() : step), m_cb(cb)
+    {}
+
+    MatView(MatView&& src) = default;
```

Comment:
Getting the idea of why it is how it is, it feels to be a big burden to actually disallow copying, since any code like below is impossible:
```
auto my_view = get_view();
step1(std::move(my_view));
step2(std::move(my_view)); 
```
And we're kinda forced to have a functional style design with `step2(step1(std::move(my_view)))` which makes sense yet might be troublesome given how OpenCV APIs are generally written.

Reply:
You can pass obtained view by reference to any function, so step1(my_view) will perfectly work if it's signature is void(cv::MatView&). MatView was designed as a short-lived entity to abstract data access, it doesn't own the memory, the only thing here is the optional callback, and non-copy behavior guarantees that it will be called only once, without involvement of shared ownership etc not to complicate things. You get the view, move it into the magazine if required (ownership is transferred) and when the view is erased from the magazine, callback is called. I'll mark copy constructors as deleted

Reply:
Ah, right. Passing by ref is always an option. Consider resolved then.

Reply:
> MatView was designed as a short-lived entity to abstract data access, it doesn't own the memory, the only thing here is the optional callback, and non-copy behavior guarantees that it will be called only once, without involvement of shared ownership etc not to complicate things.

Yep, this is totally understandable from API and documentation, so good job!